### PR TITLE
Check for duplicate Configure or ConfigureServices methods.

### PR DESCRIPTION
**Pull request by Tratcher**
#212 
@davidfowl @rausmaus @HaoK

### CODE REVIEW

**Comment by HaoK** [comment]
:shipit:

File: src/Microsoft.AspNet.Hosting/Startup/StartupLoader.cs [hunk 1]
```diff
@@ -105,8 +105,23 @@ private static MethodInfo FindMethod(Type startupType, string methodName, string
...
+            if (selectedMethods.Count == 0)
+            {
+                selectedMethods = methods.Where(method => method.Name.Equals(methodNameWithNoEnv)).ToList();
+                if (selectedMethods.Count > 1)
+                {
+                    throw new InvalidOperationException(string.Format("Having multiple overloads of method '{0}' is not supported.", methodNameWithNoEnv));
+                }
+            }
+
+            var methodInfo = selectedMethods.FirstOrDefault();
```
Comment (HaoK) [inline src/Microsoft.AspNet.Hosting/Startup/StartupLoader.cs]: nit: this should be safe to use Single() now right?

Comment (HaoK) [inline src/Microsoft.AspNet.Hosting/Startup/StartupLoader.cs]: Guess not it doesn't always throw

**Comment by StevenLiekens** [comment]
Hey sorry to bring up old issues but why not just support overloads instead of throwing?

Right now it's only possible to `Configure` everything in one method. This method gets very large, very quickly.

```csharp
public void Configure(
    IApplicationBuilder app, 
    IHostingEnvironment env,
    ICustomService1 service1,
    ICustomService2 service2,
    ICustomService3 service3,
    ICustomServiceN serviceN,
    )
{
    // build app
    app.UseThis();
    app.UseThat();
    app.Etc();
   
   // initialize custom services
   service1.ConfigureStuff(/* with lots of options */);
   
   service2.MoreStuffToConfigure(/* with lots of options */);
   
   service3.NotDoneConfiguringYet(/* ... */);
   
   serviceN.ThisIsGettingBoring(/* I think you get the point /*);
}
```

I can't think of a technical reason why I shouldn't just be able to split this into overloads?

```csharp
public void Configure( IApplicationBuilder app, IHostingEnvironment env)
{
    // build app
    app.UseThis();
    app.UseThat();
    app.Etc();
}

public void Configure(ICustomService1 service2, IHostingEnvironment env)
{
    //  // initialize custom service 1
    service1.ConfigureStuff(/* with lots of options */);
}

public void Configure(ICustomService1 service2, IHostingEnvironment env)
{
    //  // initialize custom service 2
     service2.MoreStuffToConfigure(/* with lots of options */);
}

// etc
```

The framework could just as well call every overload instead of throwing.

**Comment by Tratcher** [comment]
Comments on closed issues are not tracked, please open a new issue with the details for your scenario.

**Pull request merged**